contracts/src/v0.8/vrf/VRFCoordinatorV2Plus.sol

```diff
@@ -591,11 +591,114 @@ contract VRFCoordinatorV2Plus is VRF, TypeAndVersionInterface, SubscriptionAPI {
+  event CoordinatorDeregistered(address coordinatorAddress);
+
+  /// @notice emitted when migration to new coordinator completes successfully
+  /// @param newCoordinator coordinator address after migration
+  /// @param prevSubId old subscription ID
+  /// @param newSubId new subscription ID
+  event MigrationCompleted(
+    address newCoordinator,
+    uint64 prevSubId,
+    uint64 newSubId
+  );
+
+  /// @notice emitted when migrate() is called and given coordinator is not registered as migratable target
+  error CoordinatorNotRegistered(address coordinatorAddress);
+
+  /// @notice emitted when migrate() is called and given coordinator is registered as migratable target
+  error CoordinatorAlreadyRegistered(address coordinatorAddress);
+
+  /// @dev encapsulates data to be migrated from current coordinator
+  struct V1MigrationData {
+      uint8 fromVersion;
+      uint256 subID;
+      address subOwner;
+      address[] consumers;
+      uint96 linkBalance;
+      uint96 ethBalance;
+  }
+
+  function isTargetRegistered(address target) internal view returns (bool) {
+    for (uint256 i = 0; i < s_migrationTargets.length; i++) {
+      if (s_migrationTargets[i] == target) {
+          return true;
+      }
+    }
+    return false;
+  }
+
+  function registerMigratableCoordinator(address target) external onlyOwner {
+    if (isTargetRegistered(target)) {
+      revert CoordinatorAlreadyRegistered(target);
+    }
+    s_migrationTargets.push(target);
+    emit CoordinatorRegistered(target);
+  }
+
+  function deregisterMigratableCoordinator(address target)
+      external
+      onlyOwner
+  {
+    uint256 nTargets = s_migrationTargets.length;
+    for (uint256 i = 0; i < nTargets; i++) {
+      if (s_migrationTargets[i] == target) {
+        s_migrationTargets[i] = s_migrationTargets[nTargets - 1];
+        s_migrationTargets[nTargets - 1] = target;
+        s_migrationTargets.pop();
+        emit CoordinatorDeregistered(target);
+        return;
+      }
+    }
+    revert CoordinatorNotRegistered(target);
+  }
+
+  function migrate(uint64 subId, address newCoordinator) external {
+    if (!isTargetRegistered(newCoordinator)) {
+      revert CoordinatorNotRegistered(newCoordinator);
+    }
+    (uint96 balance, uint96 ethBalance, address owner, address[] memory consumers) = getSubscription(subId);
+    require(owner == address(this), "Not owner");
+    require(!pendingRequestExists(subId), "Pending request exists");
+
+    V1MigrationData memory migrationData = V1MigrationData({
+      fromVersion: migrationVersion(),
+      subID: subId,
+      subOwner: owner,
+      consumers: consumers,
+      linkBalance: balance,
+      ethBalance: ethBalance
+    });
+    bytes memory encodedData = abi.encode(migrationData);
+    deleteSubscription(subId);
+    uint64 newSubId = IVRFMigratableCoordinatorV2Plus(newCoordinator).onMigration(encodedData);
+    
+    require(LINK.transfer(address(newCoordinator), balance), "insufficient funds");
```

Comment:
I don't think we've been doing zero checks for transfers. e.g. see `cancelSubscriptionHelper`. If balance is zero, I think this will just be a no-op. will add test for it